File: src/main/java/io/github/takenoha/towerdefense/persistence/BlockStateSnapshot.java
package io.github.takenoha.towerdefense.persistence;

import java.util.Objects;

/** The serialised block data, block-state data, and tile payload observed in the world. */
public record BlockStateSnapshot(String blockData, String blockState, String tileNbt) {
    /** Keeps callers which only need ordinary BlockData snapshots source-compatible. */
    public BlockStateSnapshot(String blockData, String blockState) {
        this(blockData, blockState, "");
    }

    public BlockStateSnapshot {
        if (blockData == null || blockData.isBlank()) {
            throw new IllegalArgumentException("blockData must not be blank");
        }
        if (blockState == null || blockState.isBlank()) {
            throw new IllegalArgumentException("blockState must not be blank");
        }
        if (tileNbt == null) {
            throw new NullPointerException("tileNbt");
        }
        Objects.requireNonNull(blockData, "blockData");
        Objects.requireNonNull(blockState, "blockState");
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/persistence/BlockStateSnapshot.kt
package io.github.takenoha.towerdefense.persistence

import java.util.Objects
import kotlin.jvm.JvmRecord

/** The serialised block data, block-state data, and tile payload observed in the world. */
@JvmRecord
data class BlockStateSnapshot(
    val blockData: String,
    val blockState: String,
    val tileNbt: String,
) {
    /** Keeps callers which only need ordinary BlockData snapshots source-compatible. */
    constructor(blockData: String, blockState: String) : this(blockData, blockState, "")

    init {
        if (blockData.isBlank()) {
            throw IllegalArgumentException("blockData must not be blank")
        }
        if (blockState.isBlank()) {
            throw IllegalArgumentException("blockState must not be blank")
        }
        Objects.requireNonNull(tileNbt, "tileNbt")
        Objects.requireNonNull(blockData, "blockData")
        Objects.requireNonNull(blockState, "blockState")
    }
}
